File: telegram-bot/bot.js
require('dotenv').config();
const { Telegraf } = require('telegraf');

const db = require('./src/db');
const startHandler = require('./src/handlers/start');
const setpreferred = require('./src/handlers/setpreferred');
const rateHandler = require('./src/handlers/rate');
const messageHandler = require('./src/handlers/message');
const callbackQueryHandler = require('./src/handlers/callbacks');
const inlineQueryHandler = require('./src/handlers/inline');

const token = process.env.BOT_TOKEN;
if (!token) {
  console.error('Missing BOT_TOKEN. Copy .env.example to .env and fill it in.');
  process.exit(1);
}

const bot = new Telegraf(token);

bot.command('start', startHandler);
bot.command('setpreferred', setpreferred.command);
bot.command('rate', rateHandler);

bot.on('text', messageHandler);
bot.on('callback_query', callbackQueryHandler);
bot.on('inline_query', inlineQueryHandler);

bot.catch((err, ctx) => {
  console.error(`Error while handling update ${ctx.updateType}:`, err);
});

db.pruneOldInteractions();

bot.launch().then(() => {
  console.log('Bot started.');
});

process.once('SIGINT', () => bot.stop('SIGINT'));
process.once('SIGTERM', () => bot.stop('SIGTERM'));

File: telegram-bot/bot.py
import asyncio
import os

from dotenv import load_dotenv
from telethon import TelegramClient, events

from src import api, db
from src.handlers import callbacks, graph, inline, message, rate, removepreferred, setpreferred, start

load_dotenv()

API_ID = os.environ.get('API_ID')
API_HASH = os.environ.get('API_HASH')
BOT_TOKEN = os.environ.get('BOT_TOKEN')

if not API_ID or not API_HASH or not BOT_TOKEN:
    raise SystemExit(
        'Missing API_ID, API_HASH, or BOT_TOKEN. Copy .env.example to .env and fill it in '
        '(see SETUP.md for how to get API_ID/API_HASH from my.telegram.org).'
    )

client = TelegramClient('bot_session', int(API_ID), API_HASH)

client.add_event_handler(start.start_handler, events.NewMessage(pattern=r'(?i)^/start(?:@\w+)?$'))
client.add_event_handler(
    setpreferred.command, events.NewMessage(pattern=r'(?i)^/setpreferred(?:@\w+)?$')
)
client.add_event_handler(
    removepreferred.command, events.NewMessage(pattern=r'(?i)^/removepreferred(?:@\w+)?$')
)
client.add_event_handler(rate.rate_handler, events.NewMessage(pattern=r'(?i)^/rate(?:@\w+)?$'))

client.add_event_handler(
    message.message_handler,
    events.NewMessage(func=lambda e: not (e.raw_text or '').startswith('/')),
)

client.add_event_handler(callbacks.setpref_callback, events.CallbackQuery(pattern=rb'^sp:'))
client.add_event_handler(callbacks.removepref_callback, events.CallbackQuery(pattern=rb'^rp:'))
client.add_event_handler(callbacks.convert_refresh_callback, events.CallbackQuery(pattern=rb'^cv:'))
client.add_event_handler(graph.graph_callback, events.CallbackQuery(pattern=rb'^gr:'))

client.add_event_handler(inline.inline_handler, events.InlineQuery())


async def main():
    db.prune_old_interactions()
    await client.start(bot_token=BOT_TOKEN)
    print('Bot started.')
    try:
        await client.run_until_disconnected()
    finally:
        await api.close_session()


if __name__ == '__main__':
    asyncio.run(main())
